### ISSUE TITLE
Supporting CollectionView?

### ISSUE THREAD

**Issue by AAAstorga** [alexaubry/BulletinBoard#33]
### New Issue Checklist

- [ x] I updated BulletinBoard to the latest version.
- [ x] I read the [Contribution Guidelines](https://github.com/alexaubry/BulletinBoard/blob/master/CONTRIBUTING.md).
- [ x] I read the [documentation](https://alexaubry.github.io/BulletinBoard).
- [ x] I searched for [existing GitHub issues](https://github.com/alexaubry/BulletinBoard/issues).

### Issue Description

I was wondering if it is possible to support a collection view within in the bulletin page? For example, have a horizontal scrolling collection view that represents a list of images. Is this possible?

Thanks!

### Environment

- **iOS Version**: [INSERT iOS VERSION HERE]
- **Device(s)**: [INSERT DEVICE(S) HERE]

**Comment by alexaubry** [comment]
It is possible to create a custom `BulletinItem` class that contains a collection view.

## Create a wrapper for the collection view

As collections view cannot directly be used within stack views, you need to create a wrapper view to provide its size.

~~~swift
class CollectionViewWrapper: UIView {

    let collectionView: UICollectionView
    
    // change this value to the height you want for your collection view
    var pageHeight: CGFloat = 256

    init(collectionView: UICollectionView) {
        self.collectionView = collectionView
        super.init(frame: .zero)

        // pin the collection view to the edges of the wrapper
        addSubview(collectionView)
        collectionView.translatesAutoresizingMaskIntoConstraints = false
        collectionView.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
        collectionView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
        collectionView.topAnchor.constraint(equalTo: topAnchor).isActive = true
        collectionView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true

        // allow the collection view to be resized to fit the width of the bulletin 
        setContentHuggingPriority(.defaultLow, for: .horizontal)

        // do not allow the collection view to be resized vertically
        setContentCompressionResistancePriority(.required, for: .vertical)

    }

    @available(*, unavailable, message: "Use init(collectionView:)")
    required init?(coder aDecoder: NSCoder) {
        fatalError("Use init(collectionView:)")
    }

    override var intrinsicContentSize: CGSize {
        // use 0 to have a flexible width, and use the fixed height
        return CGSize(width: 0, height: pageHeight)
    }

}
~~~

## Create the Bulletin Item

Now that you can use collection view within stack views, you can implement your custom `BulletinItem` to display it.

It needs to implement the `UICollectionViewDataSource` and `UICollectionViewDelegateFlowLayout` protocols.

~~~swift

class CollectionViewBulletinItem: NSObject, BulletinItem {

    var collectionView: UICollectionView?

        func makeArrangedSubviews() -> [UIView] {

        let appearance = BulletinAppearance()
        let interfaceBuilder = BulletinInterfaceBuilder(appearance: appearance)

        var arrangedSubviews: [UIView] = []

        // Title Label

        let titleLabel = interfaceBuilder.makeTitleLabel(text: "Image")
        arrangedSubviews.append(titleLabel)

        // Collection View

        let flowLayout = UICollectionViewFlowLayout()
        flowLayout.scrollDirection = .horizontal

        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: flowLayout)
        collectionView.backgroundColor = .white

        let collectionWrapper = CollectionViewWrapper(collectionView: collectionView)

        self.collectionView = collectionView
        collectionView.register(ImageCollectionViewCell.self, forCellWithReuseIdentifier: "cell")
        collectionView.dataSource = self
        collectionView.delegate = self

        arrangedSubviews.append(collectionWrapper)

        // Action Button

        let actionButton = interfaceBuilder.makeActionButton(title: "Done")
        arrangedSubviews.append(actionButton)

        return arrangedSubviews

    }

    func tearDown() {
        collectionView?.dataSource = nil
        collectionView?.delegate = nil
    }

}

extension CollectionViewBulletinItem: UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {

    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return dataSource.numberOfImages
    }

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        return cell
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {

        // Fit the images for the height of the table view

        let image = dataSource.image(at: indexPath.row)
        let aspectRatio = image.size.width / image.size.height

        let height = collectionView.frame.height
        let width = height * aspectRatio

        return CGSize(width: width, height: height)

    }

}
~~~

> I omitted some implementation details of the collection view data source and of BulletinItem.

You can use `CollectionViewBulletinItem` as any other `BulletinObject`.

**Comment by alexaubry** [comment]
After #32 is solved, it should be easier to create custom bulletin items. You should watch this issue if you don't want to implement the action button handling manually.

**Comment by alexaubry** [comment]
I've added an example for collection view items in the latest commit, thanks for the idea 🚀

Feel free to open a new issue if you have more questions regarding collection views.
